wifinina/tcp.go

```diff
@@ -32,7 +60,97 @@ type readBuffer struct {
 	size int
 }
 
+type accessPoint struct {
+	ssid     string
+	password string
+	timeout  time.Duration
+}
+
+// HandleWifi ensures device is connected to an access point (if any) and is called before any network operation
+func (drv *Driver) HandleWifi() error {
+
+	// return fast when no access points configured
+	if drv.accessPoints == nil {
+		return nil
+	}
+
+	var activeAccessPoint *accessPoint
+	var reconnectTime time.Time
+	var stopTime time.Time
+
+	if drv.timeout != 0 {
+		stopTime = time.Now().Add(drv.timeout)
+	}
+
+	for {
+
+		// global timeout elapsed, throw
+		if !stopTime.IsZero() && time.Now().After(stopTime) {
+			return errors.New("global wifi timeout elapsed")
+		}
+
+		st, err := drv.dev.GetConnectionStatus()
```

Comment:
Apart from this PR, it is better to make sure that there is no communication with external devices here.
Since HandleWifi() is called by most methods, it is better to get out of it in as short a time as possible.

Reply:
Not sure what you propose.
`HandleWifi`s job is to ensure the device is connected to AP. It is blocking by definition, not returning until condition is satisfied, since no sense to proceed with HTTP if we have no connection.

Just describing my thinking, but open for discussion and any improvements, obviously.

Reply:
`GetConnectionStatus()`, when called, communicates with SPI and returns the result.
Since `HandleWifi()` is called by many functions, it is better to be able to know the ConnectionStatus without SPI communication.

```
func (d *Device) GetConnectionStatus() (ConnectionStatus, error) {
	status, err := d.getUint8(d.req0(CmdGetConnStatus))
	return ConnectionStatus(status), err
}
```

Reply:
🤷 It's only firmware in NINA chip knows for sure, we have to ask it. I see no alternative solution.

Reply:
I can think of optimistic approach: assume everything is OK, handle error by re-connecting to AP and repeating the failed command?

Reply:
Spent some time trying figure out any way to avoid SPI in `HandleWifi`. I don't see a way.
Thoughts:
- Every place we have `HandleWifi` function, there is always subsequent SPI call[s]. So it's not like we touch SPI when it can be completely avoided.
- HTTP communication is slow. SPI is not the bottleneck here.